### DISCUSSION TITLE
Transient users and default-roles, and the experimental state

### DISCUSSION THREAD

**Discussion by dasniko** [keycloak/keycloak#26637] [Feedback]
Hi @hmlnarik 

I'm currently testing the experimental transient users which were introduced with #23977.

So far, it works like expected, but I experienced that a transient user doesn't get the `default-roles-<realm>` role associated. Is this intended or just missing?

After a while I discovered somewhere in the docs (or just release notes?) that I have to set this mapping explicitly. After adding the `default-roles-...` in the IdP mappers as a "hardcoded role mapper", it works like a charm. But it isn't/wasn't obvious for me to do this extra step, as regular users will get assigned to the default roles automatically.

Thanks & regards!

**Comment by dasniko** [comment]
After some more more testing and playing around with this feature, I'd love to get this in _preview_ or even _supported_ state. Are there already any plans to elevate this feature? When can we expect this to happen?

Especially in EU and more especially in Germany, when working with external IdPs from government parties (like e.g. bundID (federal IdP in Germany for all citizens) or Elster (federal portal for companies/legal entities), who requires you to delete data after some time of inactivity (which is also required by the GDPR in EU), this feature would ease a lot of deployments and custom extensions around user deletion. And Keycloak is heavily used in Germany from public institutions in conjunction with bundID and Elster...!

/cc @ahus1

**Reply by ahus1** [reply]
Thank you for describing the benefits of this feature. 

Making this "preview" or even "supported" depends more on feedback in a way "I am using it already and it works well for me in the following scenarios", and less on "I want to see it to be supported before I want to use it". 

So I suggest the following: 

* Collect some show cases / describe scenarios where people are already using it. 
* Finding some volunteers to fill the gaps that exist in the documentation. There is already a documentation stub here that can be extended and then moved to the server guide. https://github.com/keycloak/keycloak/blob/main/docs/transient-users.md

**Reply by jonkoops** [reply]
@sschu might be relevant for you as well

**Reply by dasniko** [reply]
Finding customers from the public sector in Germany who are willing to test an _"experimental"_ feature is hard... even if they have in mind that also _"preview"_ features, like the token exchange, seem to never get to supported mode (yes, _I_ know there are reasons for it, but people out there in the wild don't!)

However, I've done tests with two customers in limited test environments with Keycloak 23.0. We've connected various external identity providers, with generic SAML and OIDC (no social providers!) and turned "Do not store users" on. After mapping the required and desired roles, groups and attributes, we encountered no problems.
Tests with current KC24 are still pending.

**Reply by dasniko** [reply]
IMO the documentation is already mentioning all relevant information, at least from that what I tested. I just added a little not to pay attention to the `default-roles-{realm}`, as this is not obvious.
PR: https://github.com/keycloak/keycloak/pull/28664

**Reply by dasniko** [reply]
With KC24 there seems to be a new bug in admin-ui/-api when trying to access the user details: https://github.com/keycloak/keycloak/issues/28666

**Reply by ahus1** [reply]
Update on the issue: Thank you for @hmlnarik for fixing this. We're planning to backport this fix to KC24.

**Reply by hmlnarik** [reply]
Backported in #28718, thanks for merging it, @ahus1!

**Comment by Captain-P-Goldfish** [comment]
We are currently testing this feature and so far it looks pretty good. 
We are using an `IdentityProvider` extension to test this feature. We are retrieving eID card-data from an eID-server after TR03130 and using client-scopes to manage the resulting user-info token as we need it.
From the current state I'd say the feature works great.

But we are not finished with testing yet. If we are running into some problems we will let you know :-)

**Reply by ahus1** [reply]
Thank you for this feedback, this is exactly what we are looking for!

If you find that some documentation is missing, please consider adding the necessary bits to https://github.com/keycloak/keycloak/blob/main/docs/transient-users.md

**Comment by thomasdarimont** [comment]
Instead of using hardcoded mappers, we could also have "default roles for transient users" on realm level or in identity provider  level, same could apply for groups. This would ease the configuration a bit.

**Reply by thomasdarimont** [reply]
An alternative / simple workaround is to create a group like "transient-users" and assign the default realm role to that group. Then use one hardcoded group mapper for the IdP. This will assign automatically all required roles to the transient users brokered by this IdP.

![image](https://github.com/keycloak/keycloak/assets/314690/fb55f549-78c5-4456-8a98-a54aef7f83eb)

![image](https://github.com/keycloak/keycloak/assets/314690/c3acb3bd-889d-44b4-a826-a44012923e61)

**Reply by thomasdarimont** [reply]
Perhaps that would be a good note to add to the https://github.com/keycloak/keycloak/blob/main/docs/transient-users.md docs, what do you think @hmlnarik  ?

**Reply by ahus1** [reply]
As stated below: Please clarify if you think they should be implemented before this feature is GA or after. Like everything else, it is a matter of prioritization. 

Choose your words carefully: 
* Anything that is then considered required would need to be implemented first before this feature would be able to change its state.
* Anything that you want to change after it is supported would be harder to change if it breaks existing functionality.

**Reply by thomasdarimont** [reply]
I think with the proposed group workaround having dedicated "default-roles" / "default-groups" for transient users is not necessary for GA.

**Reply by ahus1** [reply]
About the proposed documentation change: With @dasniko testing/exploring/working with the feature at the moment, he could be a reviewer for this and provide feedback.

**Reply by thomasdarimont** [reply]
Okay, then I'll send a PR for this.

**Reply by thomasdarimont** [reply]
PR is here: https://github.com/keycloak/keycloak/pull/28874

**Reply by hmlnarik** [reply]
The PR is now merged. Thanks @thomasdarimont for this update!

**Comment by thomasdarimont** [comment]
The user links in the sessions listing do not work for transient users. I suggest to disable the links for those users and make it possible to distinguish persistent users from transient users.
![image](https://github.com/keycloak/keycloak/assets/314690/9a36b1a6-db36-47ee-a2e9-87d3da537fb7)

**Reply by dasniko** [reply]
What do you mean by "do not work"?
I already filed an issue here and it has already been fixed.

**Reply by thomasdarimont** [reply]
Tested this with 24.0.3 - not with the mainline. Good to know that this was fixed already. Does the fix also includes a way to differentiate "persistent" vs. "transient" sessions in the sessions listing?

**Reply by dasniko** [reply]
There's no differentiation.
And I don't know if this is really needed from a sessions perspective. Currently, I don't see a use case here, but I might haven't elaborated it well enough.

**Reply by thomasdarimont** [reply]
I see that the user view now shows a marker "transient" for the transient user when I click on the link in the sessions list.
![image](https://github.com/keycloak/keycloak/assets/314690/bcefac4b-083e-4d6c-9d06-3c962d885393)

**Reply by Captain-P-Goldfish** [reply]
Would be nice if the marker could also be added in the list-view

**Reply by thomasdarimont** [reply]
I think having this information in the session listing could help spotting miss configurations. 

Example scenario: Perhaps you have many IdPs and you want ALL your users to be transient. Currently you would have to inspect the IdP settings or the users individually. If we would have that info in the list it would be easier to spot.

**Reply by ahus1** [reply]
@thomasdarimont - when discussing those items, please clarify if you think they should be implemented before this feature is GA or after. Like everything else, it is a matter of prioritization. 

In this thread, while it is ok to collect some future enhancement, I would like to focus on what is essential to get it to the next stage (preview or even supported).

**Reply by thomasdarimont** [reply]
I think this is quite easy to add and should be in before this goes GA.

**Reply by ahus1** [reply]
@thomasdarimont - no-one is keeping anyone from implementing it, as long as they find a reviewer. 

IMHO prioritization should be more along the following lines:

* Anything that is then considered required would need to be implemented first before this feature would be able to change its state (possibly delaying this feature)
* Anything that you want to change after it is supported would be harder to change if it breaks existing functionality.

**Reply by thomasdarimont** [reply]
I just sent a PR for showing the indicator in the list: https://github.com/keycloak/keycloak/pull/28882

**Reply by hmlnarik** [reply]
The PR is now merged. Thanks @thomasdarimont for this update!

**Comment by Captain-P-Goldfish** [comment]
The documentation looks okay from my perspective but how about adding an additional part about how to prevent user-profile-review. In case of transient users we will always have the `first broker login`-authflow to be executed and the review of the profile is somewhat disturbing at this point. So it would be nice if the documentation for this feature would shortly explain that this can be overcome by copying the first-broker-login flow and disabling the `Review Profile` step. 
I could add a PR for this or do you think this shouldn't be part of the documentation?

**Reply by hmlnarik** [reply]
A PR would be welcome. Thanks!

**Reply by Captain-P-Goldfish** [reply]
Added a PR https://github.com/keycloak/keycloak/pull/28889

**Reply by hmlnarik** [reply]
The PR is now merged. Thanks @Captain-P-Goldfish for this update!

**Comment by abidou46** [comment]
I want to see it to be supported before. Ve want to use it for the same reason described by @dasniko https://github.com/keycloak/keycloak/discussions/26637#discussioncomment-9092182

**Comment by miikahamynen-siili** [comment]
We have been doing some very light testing on this feature, and noticed that the transient users are not linked to the identity provider (see sessions -> user details -> identity provider links tab). Is that working as intended or a bug?

**Comment by dasniko** [comment]
I just tested the transient users feature with the organizations.
It works like expected, I can get the organization info in the token(s). That's good. 👍

In the admin UI, I don't see users in the organization -> members tab, but this is expected, as there are no stored users. ✔

When I access a transient user through the sessions menu, then click on the users details and go to the organizations tab, the call to the endpoint `/admin/realms/demo/organizations/members/<lightweight-user-id>/organizations` returns a `404` and leads to an error screen in the browser. I'd suggest to avoid this for transient users, either the 404 or the error screen, or both.

Can we expect a transition of this experimental feature to preview or support soon?

**Comment by stianst** [comment]
Transient users feature is most likely going to be removed. It has a number of issues like:

* Sessions may be long lived and persisted - so defeats the whole purpose really
* There is no way to identify what information is stored about a user; you can query users, but not sessions from this perspective
* Can have many duplicates for the same user (a non-transient + many transient); this is wasteful and also not great from a privacy perspective

What we should do instead is introduce some automatic user deletion policies; that can automatically delete a user with no sessions if it's a brokered user, maybe even after a certain grace period (delete brokered user after 7 days of inactivity for example).

Please see https://github.com/keycloak/keycloak/issues/38108 as a follow-up to this.

**Reply by Captain-P-Goldfish** [reply]
This would be a big problem for the german government.
We are relying on this feature because we are not allowed to persist the user data we are reading.

In germany we are currently developling several services that require the transient-users feature:
Examples are:
* [bundId](https://github.com/opdt/keycloak-extension-bundid)
* [eID](https://gitlab.opencode.de/governikus/keycloak-eid-identity-provider)
* [pid provider - eudi wallet](https://gitlab.opencode.de/governikus/eudi-wallet-pid-provider-governikus-kg)

These are some of the services but certainly not all. So basically we are using this feature as the identification of an user and not as an authentication.

**Reply by Captain-P-Goldfish** [reply]
Please tell us if you are still going to remove this feature. This is very important for us to know!

**Reply by sventorben** [reply]
@stianst Can you elaborate a bit when you are planning to remove it and what the plans are with #38198, please?
I have a customer use case where transient user feature would come in very handy and not having anything like that in place would somewhat defeat the use case.

**Reply by karisalonen** [reply]
We use (not in production yet) this feature in our implementation too, so We would be interested to hear if this is going to be removed.

**Reply by stianst** [reply]
We'll want to have https://github.com/keycloak/keycloak/issues/39888 completed before we remove transient users; I would think we'd remove that in 27, but not 100% sure at this point. @pedroigor FIY

**Reply by Captain-P-Goldfish** [reply]
thanks for sharing that @stianst. I think this is a very good alternative to the transient-users feature.

**Reply by alwibrm** [reply]
We're leveraging this feature, too.

**Comment by nvoutsin** [comment]
The transient users feature may have its shortcomings, as described by @stianst, but there are good reasons to keep it available when Keycloak is used as a thin SSO/broker layer, rather than as a full-blown IAM system with its own user database.

As already mentioned, in the EU, storing personal data in a database, even briefly and regardless of how aggressive the deletion policy may be, effectively changes the outcome of the Data Protection Impact Assessment for such deployments and can ultimately kick you out of the game.

User consent and T&C are also not sufficient workarounds. In SSO/broker scenarios, asking users for consent to store their personal data is problematic by design, since the entire purpose of the flow is to process, not store, the profile(s) of a user retrieved from remote user repositories.

While the line between a “transient user” and “temporarily stored data” can be blurry, being able to say “we do not store user data in the database” is often essential. I believe the maintainers should consider keeping the transient users feature available.

**Reply by sschu** [reply]
Do you have any source for your claim that there is a difference between storing data in a database and storing it in memory?

**Reply by nvoutsin** [reply]
from an architectural point of view, the difference is pretty clear.
from a security point of view, the difference comes from the threat models you have to deal with.
from a legal/regulatory point of view, it helps reinforce the statement “we do not store user data,”. 
Anyone who has worked with Data Protection Officers knows this can make life a lot easier. I believe others here have given similar feedback as well.

#39888, might be a good mitigation, still, there are plenty of deployment scenarios (like auth brokering) where a db with users data isn’t needed at all, and the transient-users feature is a great fit. Even advanced on-the-fly reconciliation flows can be built with transient users, at the cost of a few extra API calls.

And while transient users don’t fit every use case, my suggestion is to keep the option available for lightweight SSO/broker setups. It should be up to the implementer to decide what to use and how to handle any limitations of the feature

**Comment by dominikschlosser** [comment]
I wanted to add a use case for `transient-users` feature to this discussion, that is used heavily at my public sector customer: Using Keycloak as a transient OIDC connector to an external SAML IdP (in that case `BundID`) for the purpose of identifying users (not authenticating/logging them in). This is needed in a lot of places, where users need to present their ID card through `BundID` (and soon via the EUDI-Wallet), regardless if they are already logged in or not. This is needed as a "signature replacement" where users have to present their ID card before completion of a government application form and comes from a time where users where always physically present and needed to show their ID card in person.

So replacing this with a user deletion policy is not enough, we need to ensure that no persistent data is stored at all (like @nvoutsin already pointed out) AND that subsequent "logins" in this realm do not re-use any data from a previously linked account or trigger SSO.

So removing this would actually hurt quite a bit.

**Reply by Captain-P-Goldfish** [reply]
We also recently discussed this topic.

1. Transient users are also stored in the database for the lifetime of the session. The user data is not only held in the cache, but is also stored in the database object of the session.
2. We considered storing the data from the remote provider in the singleUseObjects cache and only storing the reference key in the user object. This would ensure that the user data is removed even if the user itself is not deleted in time.
   The major downside of this approach is that we would need to implement a custom mapper and add it by default to our clients, in the dedicated client scope, in order to include these values in tokens. This would reduce fine-grained control over token contents or require a very complex mapper implementation that will try to simulate the execution of all mappers on the stores value object.
